src/foam/nanos/notification/email/DAOResourceLoader.java

```diff
@@ -32,38 +32,59 @@
 {
   protected String groupId_;
   protected String locale_;
+  protected String spid_;
 
-  public static EmailTemplate findTemplate(X x, String name, String groupId, String locale) {
+  public static EmailTemplate findTemplate(X x, String name, String groupId, String locale, String spid) {
     DAO groupDAO = (DAO) x.get("groupDAO");
     DAO emailTemplateDAO = (DAO) x.get("localEmailTemplateDAO");
-
-    /*
-    TODO:
-
-    name  group locale spid
-      Y     Y     Y     Y
-      Y     Y     Y     *
-      Y     Y     *     Y
-      Y     Y     *     *
-      Y     *     *     Y
-      Y     *     *     *
-    */
+    EmailTemplate emailTemplate = null;
 
     do {
-      EmailTemplate emailTemplate = (EmailTemplate) emailTemplateDAO
+      emailTemplate = (EmailTemplate) emailTemplateDAO
         .find(
           AND(
-            EQ(EmailTemplate.NAME,   name),
-            EQ(EmailTemplate.GROUP,  SafetyUtil.isEmpty(groupId) ? "*" : groupId),
+            EQ(EmailTemplate.NAME, name),
+            EQ(EmailTemplate.GROUP, SafetyUtil.isEmpty(groupId) ? "*" : groupId),
+            EQ(EmailTemplate.SPID, spid),
             EQ(EmailTemplate.LOCALE, locale)
           ));
 
-      if ( emailTemplate == null ) {
+      if ( emailTemplate == null && ! SafetyUtil.isEmpty(spid) && ! SafetyUtil.isEmpty(locale) ) {
+        emailTemplate = (EmailTemplate) emailTemplateDAO
+          .find(
+            AND(
+              EQ(EmailTemplate.NAME, name),
+              EQ(EmailTemplate.SPID, spid),
+              EQ(EmailTemplate.LOCALE, locale)
+            ));
+      }
+
+      if ( emailTemplate == null && ! SafetyUtil.isEmpty(locale) ) {
+        emailTemplate = (EmailTemplate) emailTemplateDAO
+          .find(
+            AND(
+              EQ(EmailTemplate.NAME, name),
+              EQ(EmailTemplate.GROUP,  SafetyUtil.isEmpty(groupId) ? "*" : groupId),
+              EQ(EmailTemplate.LOCALE, locale)
+            ));
+      }
+
+      if ( emailTemplate == null && ! SafetyUtil.isEmpty(spid) ) {
         emailTemplate = (EmailTemplate) emailTemplateDAO
           .find(
             AND(
               EQ(EmailTemplate.NAME,  name),
-              EQ(EmailTemplate.GROUP, SafetyUtil.isEmpty(groupId) ? "*" : groupId)
+              EQ(EmailTemplate.GROUP,  SafetyUtil.isEmpty(groupId) ? "*" : groupId),
+              EQ(EmailTemplate.SPID, spid)
+            ));
+      }
+
+      if ( emailTemplate == null && ! SafetyUtil.isEmpty(spid) ) {
+        emailTemplate = (EmailTemplate) emailTemplateDAO
+          .find(
+            AND(
+              EQ(EmailTemplate.NAME,  name),
```

Comment:
I think we can't just leave values out, otherwise, we could get an unspecified match, but should instead explicitly specify the wildcard value.

Reply:
We can meet to discuss.

Reply:
@kgrgreer @Nauna  Please review again, thanks!